### ISSUE TITLE
can't build for coconut-svsm?

### ISSUE THREAD

**Issue by Unik-lif** [coconut-svsm/svsm#72]
Hello!
In fact I am not so familiar with ld. I basically followed your advice in INSTALL.md to try to build svsm.
However, it seems that the linker got some problems here.
```
link@super-server:~/svsm$ ld --version
GNU ld (GNU Binutils) 2.40
Copyright (C) 2023 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) a later version.
This program has absolutely no warranty.
link@super-server:~/snp-sgx/svsm$ rustc --version
rustc 1.73.0-nightly (28eb857b9 2023-08-12)
link@super-server:~/snp-sgx/svsm$ make
cargo build --features "default" --bin stage2
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
objcopy -O binary "target/svsm-target/"debug"/stage2" stage1/stage2.bin
cargo build --features "default" --bin svsm
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
objcopy -O elf64-x86-64 --strip-unneeded "target/svsm-target/"debug"/svsm" stage1/kernel.elf
touch stage1/svsm-fs.bin
cc -c -o stage1/stage1.o stage1/stage1.S
cc -o stage1/stage1 stage1/stage1.o stage1/reset.o -nostdlib -Wl,--build-id=none -Wl,-Tstage1/stage1.lds -fPIE
/usr/local/bin/ld: stage1/stage1.o: relocation R_X86_64_32 against `.startup.text' can not be used when making a PIE object; recompile with -fPIE
/usr/local/bin/ld: failed to set dynamic section sizes: bad value
collect2: error: ld returned 1 exit status
make: *** [Makefile:53：stage1/stage1] Error 1
```
As I add `-fPIE` in this Makefile, it still got the same error and not fixed. So I wonder if I got something wrong or it is a bug here?

Looking forward to your reply!

**Comment by 00xc** [comment]
This is a known issue with `ld`. So far we've only been able to build the SVSM with v2.39.

**Comment by Unik-lif** [comment]
Thanks. I'll test then.

**Comment by LouXiaoxuan** [comment]
Hi, I have met the same problem even my ld has been v2.39. How can I solve this issue?
```
$ ld --version
GNU ld (GNU Binutils) 2.39
Copyright (C) 2022 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) a later version.
This program has absolutely no warranty.
```

**Comment by stefano-garzarella** [comment]
@LouXiaoxuan strange `ld v2.39` should work. In the meantime you can try the containerized build, we have a script for that:
```
./scripts/container/build.sh
```

**Comment by 00xc** [comment]
@LouXiaoxuan if the containerized build does not work, could you try applying this patch?

```diff
diff --git a/Makefile b/Makefile
index 2a75262..72cdfef 100644
--- a/Makefile
+++ b/Makefile
@@ -49,7 +49,7 @@ stage1/stage1.o: stage1/stage1.S stage1/stage2.bin stage1/kernel.elf stage1/svsm
 stage1/reset.o:  stage1/reset.S stage1/meta.bin
 
 stage1/stage1: ${STAGE1_OBJS}
-       $(CC) -o $@ $(STAGE1_OBJS) -nostdlib -Wl,--build-id=none -Wl,-Tstage1/stage1.lds
+       $(CC) -o $@ $(STAGE1_OBJS) -nostdlib -Wl,--build-id=none -Wl,-Tstage1/stage1.lds -no-pie
 
 svsm.bin: stage1/stage1
        objcopy -O binary $< $@
```

**Comment by LouXiaoxuan** [comment]
Thanks for your prompt response! @stefano-garzarella @00xc I have tried both methods you advised, and they both work to build the svsm.bin successfully.

**Comment by deeglaze** [comment]
I also recommend the -no-pie on the `stage1/stage1` rule. It's a patch we've had to make while experimenting.

**Comment by deeglaze** [comment]
Fixed by #221? @Unik-lif if you can confirm, we may be able to close this issue.

**Comment by Unik-lif** [comment]
Yes, It seems to be right.
